Hardware/HDD/AbstractStorage.cs

```diff
@@ -0,0 +1,252 @@
+      ActivateSensor(sensorDiskReadRate);
+
+      sensorDiskWriteRate = new Sensor("Write Rate", 35, SensorType.Throughput, this, settings);
+      ActivateSensor(sensorDiskWriteRate);
+    }
+
+    public override HardwareType HardwareType {
+      get { return HardwareType.HDD; }
+    }
+
+    protected abstract void UpdateSensors();
+
+    internal class PerformanceValue {
+      public ulong perfValue { get; set; } = 0;
+      public ulong perfValueBase { get; set; } = 0;
+      public double result { get; set; } = 0;
+
+      public void Update(ulong v, ulong vBase) {
+        ulong diff_value = v - perfValue;
+        ulong diff_timebase = vBase - perfValueBase;
+
+        perfValue = v;
+        perfValueBase = vBase;
+        result = (100.0 / diff_timebase) * diff_value;
+
+        //sometimes it is possible that diff_value > diff_timebase
+        //limit result to 100%, this is because timing issues during read from pcie controller an latency between IO operation
+        if (result > 100)
+          result = 100;
+      }
+    }
+
+    private void UpdateStatisticsFromWmi(int driveIndex) {
+      string query = string.Format("SELECT * FROM Win32_PerfRawData_PerfDisk_PhysicalDisk Where Name LIKE \"{0}%\"", driveIndex);
+      ManagementObjectSearcher perfData = new ManagementObjectSearcher(query);
+      var data = (from ManagementObject x in perfData.Get()
+                  select x).FirstOrDefault();
+
+      if (data == null) {
+        perfData.Dispose();
+        return;
+      }
+
+      ulong value;
+      ulong value_base;
+
+      value = (ulong)data.Properties["PercentDiskWriteTime"].Value;
+      value_base = (ulong)data.Properties["PercentDiskWriteTime_Base"].Value;
+      perfWrite.Update(value, value_base);
+      sensorDiskWriteActivity.Value = (float)(perfWrite.result);
+
+      value = (ulong)data.Properties["PercentIdleTime"].Value;
+      value_base = (ulong)data.Properties["PercentIdleTime_Base"].Value;
+      perfTotal.Update(value, value_base);
+      sensorDiskTotalActivity.Value = (float)(100.0 - perfTotal.result);
+
+      ulong readRateCounter = (ulong)data.Properties["DiskReadBytesPerSec"].Value;
+      ulong readRate = readRateCounter - readRateCounterlast;
+      readRateCounterlast = readRateCounter;
+
+      ulong writeRateCounter = (ulong)data.Properties["DiskWriteBytesPerSec"].Value;
+      ulong writeRate = writeRateCounter - writeRateCounterlast;
+      writeRateCounterlast = writeRateCounter;
+
+      ulong Timestamp_PerfTime = (ulong)data.Properties["Timestamp_PerfTime"].Value;
+      ulong Frequency_Perftime = value_base = (ulong)data.Properties["Frequency_Perftime"].Value;
+      double current_time = (double)Timestamp_PerfTime / Frequency_Perftime;
+
+      double timedeltaSeconds = current_time - lastTime;
+      lastTime = current_time;
+      if (timedeltaSeconds > 0.2 && timedeltaSeconds < 5.0) {
```

Comment:
I'm testing out this PR from a library perspective and the only issue I'm running into /nvme/0/throughput/{34,35} lack values (read and write speed). It's because of this line: ensuring the interval from the last check was under 5 seconds. This puts an artificial constraint that all updates must be under 5 seconds in order to receive nvme read and write speed.

I wonder if it wouldn't be better to skip setting values if lastTime is zero

Also IMO, last_time should accumulate if timeDeltaSeconds is below 0.2 seconds (as if someone polled every 100ms then read and write speed will always lack a value as well)

I found the following code fixed my use case.

```csharp
double timedeltaSeconds = current_time - lastTime;
if (lastTime != 0 && timedeltaSeconds > 0.2) {
  double writeSpeed = (double)writeRate;
  writeSpeed *= (1 / timedeltaSeconds);
  sensorDiskWriteRate.Value = (float)writeSpeed;

  double readSpeed = (double)readRate;
  readSpeed *= (1 / timedeltaSeconds);
  sensorDiskReadRate.Value = (float)readSpeed;
}

if (lastTime == 0 || timedeltaSeconds > 0.2) {
  lastTime = current_time;
}
```